### ISSUE TITLE
[ADDITION] Catalyst Web Browser

### ISSUE THREAD

**Issue by JaydenDev** [Lissy93/awesome-privacy#26]
### Category

Essentials/Private Browsers

### Software to be added

All sources of the Catalyst project are proudly available [here](/JaydenDev/Catalyst). I am constantly patching bugs, and most importantly, updating dependencies to keep a safe and secure browser. The best part, absolutely zero data collection. I haven't yet mentioned, but Catalyst is built with Electron. Making it impossible to block sites from collecting data, although some methods may not work in the environment anyways. But, if you use the browser (no data collection) and use sites that respect your privacy, you'll be practically invisible! I would add a fake user agent, but that would break several sites. The browser is also stable, fast, and does one thing and does it well (be a browser...). In upcoming releases there will be a one time warning on non-private sites! Anyways, you are not obliged to add this to the list. Simply a suggestion!

### Supporting Material

[Sources](/JaydenDev/Catalyst)
[Website](https://getcatalyt.eu.org)

### Association Disclosure

I own the repository, browser, and its sources. And my code makes up about 75 percent.

### Would you like to submit a PR?

Maybe?

### Please tick the boxes

- [X] You have filled out this form accurately, and to the best of your knowledge
- [X] You have indicated weather or not you are associated with the suggested addition
- [X] A similar submission has not already [been opened](https://github.com/Lissy93/awesome-privacy/issues?q=is%3Aissue) for this software/ service
- [X] You agree to the [code of conduct](https://github.com/Lissy93/awesome-privacy/blob/master/.github/CODE_OF_CONDUCT.md)

**Comment by JaydenDev** [comment]
You are entirely free to view, download, and inspect the sources to further verify the goals and lack of data collection!

**Comment by Lissy93** [comment]
Thanks @JaydenDev , just looking into it now...
Is https://getcatalyt.eu.org/ down for you, or just me?

**Comment by Lissy93** [comment]
Your project does look cool :)

But I've had a quick look through the code, and I think that for now it doesn't quite meet the criteria. 

The biggest issue is just with Electron. There's been a fair few security issues with Electron's webview over the years, and I would guess more to come.

Btw, I might have just missed it, but under [`webPreferences`](https://github.com/JaydenDev/Catalyst/blob/master/main/main.js#L18) there's a few config options you should definitely be applying for security. Like enable sandboxing, enable context isolation, disable remote module, disable node integration, etc. Without those the user would be very vulnerable to anything malicious on any site they visit. Tho still not perfect, take a look at [minbrowser's implementation](https://github.com/minbrowser/min/blob/master/js/webviews.js).

Catalyst seems to be in the early stages of development still, not quite stable or feature-complete. There is some incomplete code within the repo, and a lot of commented out code, and hard-coded values. I advice putting in some structure, writing a few unit tests and adding linting and also maybe prettier or something for neatness.

I'd be happy to consider Catalyst again, once it's more mature, stable and with a USP.
Best of luck with it :)

**Comment by JaydenDev** [comment]
Catalyst has been in stable stages since the first major release (v1.0.0). So I don't understand how Catalyst has not matured into a basic usable browser yet. I will add the security options to webPreferences though. Maybe Catalyst will be considered mature in '4.0.0'.

**Comment by Lissy93** [comment]
In case it helps, these were the biggest security issues which jumped out at me:

- `src/functions.js` - Avoid Math.random, especially in this conext. As it relies on a weak pseudorandom number generator. You can use [SubtleCrypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) instead
- ` src/welcome.html` - Don't load resources from external sources without verifying their integreity (e.g. jQuery, tailwind etc). You can do this with [Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)
- `main/main.js` - With the webview, you must properly isolate content, setup sandboxing, partitioning, disable remote module, etc. You can learn more in the [Electron webview](https://www.electronjs.org/docs/latest/api/webview-tag) docs

---

And some general things:

Consider implement unit tests, code quality checks, type checking, dependency monitoring. Add branch rules so that no code can be merged to production without passing those checks, and being reviewed. GH Actions is great for that.

Never manipulate the dom using any dynamic values without sanitizing them. It's trivial for an attacker to craft some malicious code to snatch up user info, redirect them, pollute user space, etc.

Try to keep all your dependencies managed centrally, you've got a mix of CDN (jQuery), locally downlaoed stuff (sortable.js) devDependencies and dependencies. You're not going to be getting security / dependabot alerts for anything not in your package.json / yarn.lock.

Finally, a general issue with the way that Electron's webview is actually implemented, is as an [Out-of-Process iframe](https://www.chromium.org/developers/design-documents/oop-iframes/), these work great when used for a single domain, but if not used correctly, can be susceptible to a series of injection, buffer overflow and XXS attacks. Take great care.
And from a privacy perspective, you're going to be much more trackable, and have far fewer security features at your disposal.

---

Hope that helps :)

**Comment by JaydenDev** [comment]
The last vulnerability regarding out of Out-Of-Process iframes is impossible to fix without a total rewrite of the core tab component of the browser.

I cannot sign releases, as that requires paying Microsoft and Apple HUNDREDS of dollars for a signature thingy. This is a project written by someone who... really can't pay that.

Modifying the DOM with dynamic values.... I have to do that? Read the code, its impossible to not do so.

I can easily not use CDNs as a whole, considering if they are on a cdn they probably are also on npm.

In general, you are really asking too much of this software. You expect it to be entirely impenetrable. Which is impossible, I can add the isolation and stuff to the Electron core though.

**Comment by JaydenDev** [comment]
As well Catalyst was __never__ intended to be a browser used by anyone. Simply an Electron experiment, learning Electron. And as of right now, Catalyst isn't even highest priority. Anyway, this issue can stay closed!

**Comment by jdev082** [comment]
I'm currently looking into cracking down on various privacy issues that are present. I recently fixed an issue where bookmark icons were pulled from a Google API. I will be applying some of the aforementioned webPreferences options soon, and they will be included in the Christmas Update. There is no longer any remote content loaded. 

As for release signing, as far as I know, signing the MacOS and Windows releases could be pretty expensive. Would providing sha256 hashes suffice?

Checks. I have recently setup dependabot and vulnerability checking. As well as eslint. I've also been testing releases more thoroughly as Catalyst has become available in many software repositories, it's becoming tedious to push out releases. 

There is a library included in the Christmas Update for the snow. But this will be immediately removed in the next release, with few updates released for this library before then.

DOM Manipulation, There is indeed a lot of it. Upon searching for 'innerHTML', I found a vulnerability front and center. When bookmarks are added to the bar, the icon and title current webviews' icon and title (which is set by the page developer) are added to the HTML. Allowing for XSS. This should be easily fixable by creating the elements individually and pushing them instead of doing it all in one go like this. I will absolutely be fixing this shortly. 

I do have some questions.

When initially looking at the `generateHashkey` function I can see how one would assume this is used for security-related purposes. However, it's simply to give each tab and webview an ID that can be used for when switching between them. So I don't think true randomness is important here, but I may be wrong.

If any more privacy and security issues are to be found, they will be added to the [tracker](https://github.com/orgs/CatalystDevOrg/discussions/557) and fixed quickly.
